### PR TITLE
General macro for supporting only a subset of fields in the Backend.

### PR DESCRIPTION
**Pull request by lvella**

### CODE REVIEW

**Comment by Schaeff** [comment]
My general stance is to avoid unsafe code if possible so here's my proposal:

- Introduce a `Columns` enum which holds `Vec<String, Vec<T>>` for all possible `T` we support
- Introduce a `ProgramWithFixed` enum which holds `(Analysis<T>, Vec<String, Vec<T>>)` for all possible `T` we support
- Pass this instead of the first two arguments in the `BackendFactory` trait. The `T` generic parameter disappears.
- Now in each backend, each method end up containing a big match statement with all fields. Have the backend implement the ones it supports. This can use a macro:

```rust
impl Backend for Plonky3 {
   fn prove(&self, witness: &Columns) -> Result<Proof, Error> {
      call_with_all_fields!(Plonky3FieldElement, p3_prove, self, witness)
   }
}
```

where the macro expands to 

```rust
match self.program {
   GoldilocksProgram(p) => Ok(p3_prove(self, witness)?),
   BabybearProgram(p) => Ok(p3_prove(self, witness)?),
   _ => Err(Error::UnsupportedField)
}
```

where the implemented arms correspond to the fields which implement `Plonky3FieldElement`. If this is not something we have access to, maybe we can make use of the `KnownField` enum.

This proposal may need a bit of adjustment but from my perspective:
Cons:
- Requires changing the backend API
Pros:
- No unsafe code
- We'd have to make such a change anyway to allow backends to accept a program which internally uses different fields, think recursion, folding etc

**Comment by lvella** [comment]
I don't get the Pro number two. A change in this direction might be needed, but it is not sufficient, right? The backend interface would have to change in other aspects, and probably our structs representing the program, too.

Anyway, one pro for the current PR is that it is small and reuses the interfaces we already have. And the unsafe is contained within a small block inside the match choice that guarantees its safety at runtime.

File: backend/src/field_filter.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
+/// Creates a new `pub(crate)` general Factory struct from the RestrictedFactory
+/// struct.
+///
+/// If the type generic type of the general factory is one of the listed
```
[the remark this answers is not in the document]
Comment (leonardoalt) [reply backend/src/field_filter.rs:4]: I just meant the sentence sounds a bit weird

[the remark this answers is not in the document]
Comment (lvella) [reply backend/src/field_filter.rs]: Yeah, fixed.

File: plonky3/src/circuit_builder.rs [hunk 2]
```diff
@@ -11,6 +11,7 @@ use std::collections::{BTreeMap, HashSet};
 use crate::params::{Commitment, FieldElementMap, Plonky3Field, ProverData};
 use p3_air::{Air, AirBuilder, AirBuilderWithPublicValues, BaseAir, PairBuilder};
 use p3_matrix::{dense::RowMajorMatrix, Matrix};
+// use p3_mds::MdsPermutation;
```
Comment (leonardoalt) [inline plonky3/src/circuit_builder.rs:14]: leftover

Comment (lvella) [reply plonky3/src/circuit_builder.rs]: fixed

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**